app/src/main/java/fr/neamar/kiss/forwarder/Widgets.java

```diff
@@ -93,19 +106,65 @@ void onActivityResult(int requestCode, int resultCode, Intent data) {
+            boolean textMode = prefs.getBoolean("widget-text-mode", false);
+            if (!textMode) {
+                // request widget picker, a selection will lead to a call of onActivityResult
+                int appWidgetId = mAppWidgetHost.allocateAppWidgetId();
+                Intent pickIntent = new Intent("android.appwidget.action.APPWIDGET_PICK");
+                pickIntent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId);
+                mainActivity.startActivityForResult(pickIntent, REQUEST_APPWIDGET_PICKED);
+                return true;
+            } else {
+                // request widget picker via custom text selection
+                List<AppWidgetProviderInfo> widgetList = mAppWidgetManager.getInstalledProviders();
+                ArrayList<String> names = new ArrayList<String>();
+                ArrayList<AppWidgetProviderInfo> providers = new ArrayList<AppWidgetProviderInfo>();
+                // get all provides and names
+                for (AppWidgetProviderInfo info : widgetList) {
+                    names.add(info.provider.getClassName());
+                    providers.add(info);
+                }
+                final Widgets t = this;
+                // Prompt user to pick a widget based on their class names
+                new AlertDialog.Builder(mainActivity).setTitle("Pick a Widget").setItems(names.toArray(new String[]{}),
+                        new DialogInterface.OnClickListener() {
+                            @RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN)
+                            @Override
+                            public void onClick(DialogInterface dialog, int which) {
+                                AppWidgetProviderInfo provider = providers.get(which);
+                                if (provider != null) {
+                                    int id = mAppWidgetHost.allocateAppWidgetId();
+                                    // This requires min api 16
+                                    boolean success = mAppWidgetManager.bindAppWidgetIdIfAllowed(id, provider.provider);
+                                    Log.d("Widgets", "suc" + success);
+                                    if (success) {
+                                        // directly call onActivityResult with the selected/newly-created id.
+                                        t.onActivityResult(REQUEST_APPWIDGET_PICKED, Activity.RESULT_OK, new Intent().putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, id));
+                                    } else {
+                                        List<AppWidgetProviderInfo> infos = mAppWidgetManager.getInstalledProviders();
+                                        Intent intent = new Intent(AppWidgetManager.ACTION_APPWIDGET_BIND);
+                                        intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, id);
+                                        intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_PROVIDER, infos.get(0).provider);
```

Comment:
EXTRA_APPWIDGET_PROVIDER should be the one selected by the user, not the first one from the list.
You already have the provider `AppWidgetProviderInfo provider = providers.get(which);`

Reply:
Whoops, thanks, that's a debugging relict.